### ISSUE TITLE
00_Common_Setup.py

### ISSUE THREAD

**Issue by Curiosity-Mars** [Curiosity-Mars/rarp-psa-trajectory-reproducibility#1]
"""Shared configuration for the postoperative PSA reproducibility code.

Run this file once before a figure- or table-specific script.  The public
repository contains analysis code only; patient-level data are not distributed.
Place the two source CSV files in DATA_DIR or set DMODULE_DATA_DIR.
"""

from __future__ import annotations

import hashlib
import json
import os
import platform
import re
from pathlib import Path

import matplotlib
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import scipy
import sklearn

# Mount Google Drive when the scripts are pasted into a fresh Colab runtime.
if Path("/content").exists() and not Path("/content/drive/MyDrive").exists():
    try:
        from google.colab import drive  # type: ignore

        drive.mount("/content/drive")
    except ImportError:
        pass

# -----------------------------------------------------------------------------
# Reproducibility settings
# -----------------------------------------------------------------------------
RANDOM_STATE = 42
KMEANS_N_INIT = 10
N_BOOTSTRAP = 2000
N_CV_SPLITS = 5

SOURCE_COHORT_N = 1685
ANALYSIS_READY_N = 1455
TRAJECTORY_COHORT_N = 1434
EXPECTED_CLUSTER_COUNTS = {0: 78, 1: 984, 2: 372}

PSA_MONTHS = (1, 3, 6, 9, 12, 15, 18, 24, 30, 36, 48, 60, 72, 84, 96, 108, 120)
REFERENCE_SIGMA = 1.0
REFERENCE_K = 3

EXPECTED_ANALYSIS_SHA256 = (
    "750b04b1542d7dfd22ef3219c2e0331b5ef988cc6dbeeb85f49d39d1a13d944c"
)

def _default_root() -> Path:
    """Use Google Drive in Colab and the current directory elsewhere."""
    if Path("/content/drive/MyDrive").exists():
        return Path("/content/drive/MyDrive")
    return Path.cwd()

DATA_DIR = Path(os.environ.get("DMODULE_DATA_DIR", _default_root())).expanduser().resolve()
OUTPUT_DIR = Path(
    os.environ.get("DMODULE_OUTPUT_DIR", DATA_DIR / "Dmodule_PSA_reproducibility")
).expanduser().resolve()
OUTPUT_DIR.mkdir(parents=True, exist_ok=True)

SOURCE_COHORT_FILE = DATA_DIR / "RARP2024JulyCOMM.csv"
ANALYSIS_READY_FILE = DATA_DIR / "PSA_20250404.csv"

# -----------------------------------------------------------------------------
# Input and output helpers
# -----------------------------------------------------------------------------
def sha256_file(path: Path, chunk_size: int = 1024 * 1024) -> str:
    digest = hashlib.sha256()
    with path.open("rb") as handle:
        for chunk in iter(lambda: handle.read(chunk_size), b""):
            digest.update(chunk)
    return digest.hexdigest()

def read_csv_robust(path: Path) -> pd.DataFrame:
    """Read a CSV using encodings used by the study exports."""
    if not path.exists():
        raise FileNotFoundError(f"Required input file not found: {path}")
    last_error: Exception | None = None
    for encoding in ("utf-8-sig", "utf-8", "cp932"):
        try:
            return pd.read_csv(path, encoding=encoding, low_memory=False)
        except UnicodeDecodeError as exc:
            last_error = exc
    raise RuntimeError(f"Could not decode {path}") from last_error

def verify_rows(frame: pd.DataFrame, expected_n: int, label: str) -> None:
    if len(frame) != expected_n:
        raise ValueError(f"{label}: expected {expected_n:,} rows; found {len(frame):,}")

def verify_sha256(path: Path, expected: str) -> str:
    actual = sha256_file(path)
    if actual != expected:
        raise ValueError(
            "Input SHA-256 mismatch.\n"
            f"Expected: {expected}\n"
            f"Actual:   {actual}\n"
            f"File:     {path}"
        )
    return actual

def add_analysis_row_id(frame: pd.DataFrame) -> pd.DataFrame:
    """Add the row-order identifier used only for traceability and joins."""
    result = frame.copy()
    result["Analysis_Row_ID"] = np.arange(1, len(result) + 1, dtype=int)
    return result

def resolve_column(frame: pd.DataFrame, aliases: tuple[str, ...], role: str) -> str:
    """Resolve one source column without silently accepting ambiguity."""
    exact = [name for name in aliases if name in frame.columns]
    if len(exact) == 1:
        return exact[0]
    if len(exact) > 1:
        raise KeyError(f"Ambiguous columns for {role}: {exact}")

    normalized = {re.sub(r"\s+", "", str(c)).lower(): c for c in frame.columns}
    matches = []
    for alias in aliases:
        key = re.sub(r"\s+", "", alias).lower()
        if key in normalized:
            matches.append(normalized[key])
    matches = list(dict.fromkeys(matches))
    if len(matches) == 1:
        return matches[0]
    raise KeyError(
        f"Could not uniquely resolve {role}. Tried {aliases}. "
        "Update only the alias list after checking the original data dictionary."
    )

def psa_columns_in_time_order(frame: pd.DataFrame) -> list[str]:
    expected = [f"PSA POM{month}" for month in PSA_MONTHS]
    missing = [column for column in expected if column not in frame.columns]
    if missing:
        raise KeyError(f"Missing postoperative PSA columns: {missing}")
    return expected

def save_manifest(path: Path, payload: dict) -> None:
    with path.open("w", encoding="utf-8") as handle:
        json.dump(payload, handle, ensure_ascii=False, indent=2)

def format_p_value(value: float) -> str:
    if not np.isfinite(value):
        return "—"
    return "<0.001" if value < 0.001 else f"{value:.3f}"

def software_versions() -> dict[str, str]:
    return {
        "python": platform.python_version(),
        "numpy": np.__version__,
        "pandas": pd.__version__,
        "scipy": scipy.__version__,
        "scikit_learn": sklearn.__version__,
        "matplotlib": matplotlib.__version__,
    }

np.random.seed(RANDOM_STATE)
plt.rcParams.update(
    {
        "figure.dpi": 120,
        "savefig.dpi": 300,
        "font.family": "DejaVu Sans",
        "axes.spines.top": False,
        "axes.spines.right": False,
    }
)

print("D-module PSA reproducibility setup loaded")
print("Data directory:", DATA_DIR)
print("Output directory:", OUTPUT_DIR)
print("Random state:", RANDOM_STATE)
print("Software:", software_versions())
